File: scripts/lib/embed-section.ts
/**
 * Generic embed section building for social media platforms.
 *
 * Uses a factory function (higher-order function) to eliminate the
 * duplication of buildTweetSection, buildBlueskySection, buildMastodonSection.
 *
 * This is an application of the Strategy pattern expressed functionally:
 * the section header is a parameter, and the build logic is shared.
 *
 * @module embed-section
 */

import fs from "node:fs";

import {
  TWEET_SECTION_HEADER,
  BLUESKY_SECTION_HEADER,
  MASTODON_SECTION_HEADER,
} from "./types.ts";

// --- Generic Section Builder (Factory / Higher-Order Function) ---

/**
 * Create a section builder for a given header.
 *
 * Returns a pure function: (existingContent, embedHtml) → sectionString.
 *
 * This is a higher-order function (returns a function), eliminating the
 * duplication of three nearly-identical build functions.
 */
export const createSectionBuilder = (header: string) =>
  (existingContent: string, embedHtml: string): string => {
    const separator = existingContent.endsWith("\n") ? "\n" : "\n\n";
    return `${separator}${header}  \n${embedHtml}`;
  };

/**
 * Create a section appender for a given header.
 *
 * Returns an effectful function that appends the section to a file,
 * with idempotency: skips if the section already exists.
 */
export const createSectionAppender = (header: string) => {
  const buildSection = createSectionBuilder(header);

  return (filePath: string, embedHtml: string): void => {
    const content = fs.readFileSync(filePath, "utf-8");

    if (content.includes(header)) {
      console.log(`${header} already exists, skipping update`);
      return;
    }

    fs.writeFileSync(filePath, content + buildSection(content, embedHtml), "utf-8");
  };
};

// --- Platform-Specific Instances ---

export const buildTweetSection = createSectionBuilder(TWEET_SECTION_HEADER);
export const buildBlueskySection = createSectionBuilder(BLUESKY_SECTION_HEADER);
export const buildMastodonSection = createSectionBuilder(MASTODON_SECTION_HEADER);

export const appendTweetSection = createSectionAppender(TWEET_SECTION_HEADER);
export const appendBlueskySection = createSectionAppender(BLUESKY_SECTION_HEADER);
export const appendMastodonSection = createSectionAppender(MASTODON_SECTION_HEADER);

File: haskell/src/Automation/EmbedSection.hs
module Automation.EmbedSection
  ( EmbedSection (..)
  , createSectionBuilder
  , createSectionAppender
  , buildTweetSection
  , buildBlueskySection
  , buildMastodonSection
  , appendTweetSection
  , appendBlueskySection
  , appendMastodonSection
  ) where

import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.IO as TIO

import qualified Automation.Platforms.Bluesky as Bluesky
import qualified Automation.Platforms.Mastodon as Mastodon
import qualified Automation.Platforms.Twitter as Twitter

data EmbedSection = EmbedSection
  { esHeader :: Text
  , esEmbedHtml :: Text
  , esBuildSection :: Text -> Text -> Text
  }

createSectionBuilder :: Text -> Text -> Text -> Text
createSectionBuilder header existingContent embedHtml =
  let separator = if T.isSuffixOf "\n" existingContent then "\n" else "\n\n"
  in separator <> header <> "  \n" <> embedHtml

createSectionAppender :: Text -> FilePath -> Text -> IO ()
createSectionAppender header filePath embedHtml = do
  content <- TIO.readFile filePath
  case T.isInfixOf header content of
    True  -> putStrLn $ T.unpack header <> " already exists, skipping update"
    False -> TIO.writeFile filePath (content <> createSectionBuilder header content embedHtml)

buildTweetSection :: Text -> Text -> Text
buildTweetSection = createSectionBuilder Twitter.sectionHeader

buildBlueskySection :: Text -> Text -> Text
buildBlueskySection = createSectionBuilder Bluesky.sectionHeader

buildMastodonSection :: Text -> Text -> Text
buildMastodonSection = createSectionBuilder Mastodon.sectionHeader

appendTweetSection :: FilePath -> Text -> IO ()
appendTweetSection = createSectionAppender Twitter.sectionHeader

appendBlueskySection :: FilePath -> Text -> IO ()
appendBlueskySection = createSectionAppender Bluesky.sectionHeader

appendMastodonSection :: FilePath -> Text -> IO ()
appendMastodonSection = createSectionAppender Mastodon.sectionHeader
